### PR TITLE
Defer loading fix models

### PR DESCRIPTION
**Pull request by yongyan-gh**
Instead loading fix models in SarifErrorListItem constructor, populate the fix models whenever the VS editor loads suggested actions (fixes).

related to #304

### CODE REVIEW

File: src/Sarif.Viewer.VisualStudio.UnitTests/ErrorList/SarifErrorListItemTests.cs [hunk 1]
```diff
@@ -255,7 +255,7 @@ public void SarifErrorListItem_WhenFixHasRelativePath_UsesThatPath()
             };
 
             SarifErrorListItem item = MakeErrorListItem(result);
-
+            item.PopulateFixModelsIfNot();
```
Comment (eddynaka) [inline src/Sarif.Viewer.VisualStudio.UnitTests/ErrorList/SarifErrorListItemTests.cs:258]: >PopulateFixModelsIfNot [](http://example.com/codeflow?start=17&length=22)

you could check if you have fixes before and after.
i would expect that if we don't call the method, the fix wouldn't be loaded

File: src/Sarif.Viewer.VisualStudio.Core/Sarif/ArtifactChange.Extensions.cs [hunk 2]
```diff
@@ -41,5 +41,23 @@ public static ArtifactChangeModel ToArtifactChangeModel(this ArtifactChange file
 
             return model;
         }
+
+        public static ArtifactChangeModel ToArtifactChangeModel(this ArtifactChange fileChange, IDictionary<string, Uri> originalUriBaseIds, FileRegionsCache fileRegionsCache)
```
Comment (eddynaka) [inline src/Sarif.Viewer.VisualStudio.Core/Sarif/ArtifactChange.Extensions.cs:45]: >ToArtifactChangeModel [](http://example.com/codeflow?start=42&length=21)

pls, add some tests to cover this change

File: src/Sarif.Viewer.VisualStudio.Core/Sarif/Fix.Extensions.cs [hunk 3]
```diff
@@ -30,6 +30,23 @@ public static FixModel ToFixModel(this Fix fix, IDictionary<string, ArtifactLoca
             return model;
         }
 
+        public static FixModel ToFixModel(this Fix fix, IDictionary<string, Uri> originalUriBaseIds, FileRegionsCache fileRegionsCache)
+        {
+            fix = fix ?? throw new ArgumentNullException(nameof(fix));
+
```
Comment (eddynaka) [inline src/Sarif.Viewer.VisualStudio.Core/Sarif/Fix.Extensions.cs:36]: pls test:
1. null fix
2. fix.artifactchanges = null

File: src/Sarif.Viewer.VisualStudio.Core/Models/SarifErrorListItem.cs [hunk 4]
```diff
@@ -150,14 +150,7 @@ public SarifErrorListItem(Run run, int runIndex, Result result, string logFilePa
...
 
-            if (result.Fixes != null)
-            {
-                foreach (Fix fix in result.Fixes)
-                {
-                    var fixModel = fix.ToFixModel(run.OriginalUriBaseIds, FileRegionsCache.Instance);
-                    this.Fixes.Add(fixModel);
-                }
-            }
+            // fixes are not populated in constructor.
```
Comment (eddynaka) [inline src/Sarif.Viewer.VisualStudio.Core/Models/SarifErrorListItem.cs:153]: >// fixes are not populated in constructor [](http://example.com/codeflow?start=12&length=41)

this comment should not exist.
this comment should be a codeflow comment

File: src/Sarif.Viewer.VisualStudio.Core/Sarif/ArtifactChange.Extensions.cs [hunk 5]
```diff
@@ -41,5 +41,28 @@ public static ArtifactChangeModel ToArtifactChangeModel(this ArtifactChange file
...
+                // convert IDictionary<string, Uri> to IDictionary<string, ArtifactLocation> which
+                // is needed by Sarif.SDK extension method ArtifactLocation.TryReconstructAbsoluteUri()
+                uriToArtifactLocationMap = new Dictionary<string, ArtifactLocation>(originalUriBaseIds.Count);
+                foreach (KeyValuePair<string, Uri> entry in originalUriBaseIds)
+                {
+                    uriToArtifactLocationMap.Add(entry.Key, new ArtifactLocation { Uri = entry.Value });
+                }
+            }
+
+            return fileChange.ToArtifactChangeModel(uriToArtifactLocationMap, fileRegionsCache);
```
Comment (eddynaka) [inline src/Sarif.Viewer.VisualStudio.Core/Sarif/ArtifactChange.Extensions.cs:65]: >fileChange.ToArtifactChangeModel(uriToArtifactLocationMap, fileRegionsCache) [](http://example.com/codeflow?start=19&length=76)

there is a possibility where originalUriBaseIds == null, with that, uriToArtifactLocationMap would be null.

Can you check if fileChange.ToArtifactChangeModel won't throw an exception, pls?

Comment (yongyan-gh) [reply src/Sarif.Viewer.VisualStudio.Core/Sarif/ArtifactChange.Extensions.cs:65]: yes fileChange.ToArtifactChangeModel() allows null Dictionary<string, ArtifactLocation> as parameter and code handles that and will not throw exception.

so the change here allows null Dictionary<string, Uri> as parameter of the overload method, only add values if originalUriBaseIds is not null.

**Review by eddynaka — APPROVED** [review]
:shipit:

**Pull request merged**